text/0000-new-test-waiters.md

```diff
@@ -0,0 +1,238 @@
+
+### Using the `TestWaiter` class
+
+After building a test waiter, most users interact with a limited set of methods within this class, namely `beingAsync` and `endAsync`.
+
+The API used to signal whether an asynchronous operation has begun and ultimately ended is through the **_paired_** calls of `beginAsync` and `endAsync`: begin to denote the start of the asynchronous operation, and end to denote the end. Unique instances of async operations are identified using a `token` returned from `beginAsync`, which is subsequently provided to the `endAsync` call.
+
+To annotate the example provided above:
+
+```js
+import Component from '@ember/component';
+import { buildWaiter } from 'ember-test-waiters';
+
+// Creates a test waiter with the name 'friend-waiter' that
+// is usable by all instances of the `Friendz` component.
+let waiter = buildWaiter('friend-waiter');
+
+export default class Friendz extends Component {
+  didInsertElement() {
+    // Alerts the test waiter system that an async operation has started,
+    // storing the resulting unique token to be used to notify the test
+    // waiter system that the operation has ended.
+    let token = waiter.beginAsync();
+
+    someAsyncWork()
+      .then(() => {
+        //... some work
+      })
+      .finally(() => {
+        // Notifies the test waiter system that
+        // this unique async operation has ended.
+        waiter.endAsync(token);
+      });
+  }
+}
+```
+
+### `waitForPromise`
+
+The `waitForPromise` utility provides a convenience wrapper around the `TestWaiter` class for use with promises. It ensures the `endAsync` call is invoked in the `finally` of the configured promise.
+
+```js
+import Component from '@ember/component';
+import { waitForPromise } from 'ember-test-waiters';
+
+export default class MoreFriendz extends Component {
+  didInsertElement() {
+    waitForPromise(someAsyncWork).then(() => {
+      doOtherThings();
+    });
+  }
+}
+```
+
+This new test waiters system has been through multiple iterations of refinement, and is in use and integrated with the test isolation validation system.
+
+## Rename of `ember-test-waiters` to `@ember/test-waiters`
+
+We should consider renaming the `ember-test-waiters` repository to `@ember/test-waiters`, which would relocate it to the Ember org and scope it more explicitly.
+
+## Compatibility
+
+The `ember-test-waiters` addon is backwards compatible with the old test waiters system, allowing applications and addons to gradually migrate to using the new system.
+
+Specifically:
+
+- `registerWaiter` continues to work, and is not deprecated
+- addons using `ember-test-waiters` work _even if consumed in applications that do not use a new enough `ember-test-helpers` version_ (they won't get additional output via test isolation validation such as test waiter names or stack traces)
+
+The old test waiters system ultimately should be deprecated in its own deprecation RFC.
+
+# How We Teach This
```

Comment:
I think the "How we teach this" should include guidance for how we communicate to addon developers vs. app developers.

For example we need to communicate to addon developers that they should expose waited async and avoid passing the burden of using the wrappers to the end user.

Reply:
This is a great callout. I've added a specific section that indicates that addon authors should have further direction on the importance of adding test waiters to their addon's unmanaged async.